src/server/replica.cc

```diff
@@ -550,40 +581,108 @@ error_code Replica::StartAsDflyFlow() {
   if (flow_directive == "FULL") {
     eof_token = ToSV(resp_args_[1].GetBuf());
   } else {
-    LOG(ERROR) << "Bad FLOW response " << ToSV(io_buf->InputBuffer());
+    LOG(ERROR) << "Bad FLOW response " << ToSV(leftover_buf_->InputBuffer());
   }
-  io_buf->ConsumeInput(consumed);
+  leftover_buf_->ConsumeInput(consumed);
 
   state_mask_ = R_ENABLED | R_TCP_CONNECTED;
 
   // We can not discard io_buf because it may contain data
   // besides the response we parsed. Therefore we pass it further to ReplicateDFFb.
-  sync_fb_ =
-      ::boost::fibers::fiber(&Replica::FullSyncDflyFb, this, std::move(io_buf), move(eof_token));
+  sync_fb_ = ::boost::fibers::fiber(&Replica::FullSyncDflyFb, this, sb, move(eof_token));
 
   return error_code{};
 }
 
-void Replica::FullSyncDflyFb(unique_ptr<base::IoBuf> io_buf, string eof_token) {
+error_code Replica::StartStableSyncFlow() {
+  DCHECK(!master_context_.master_repl_id.empty() && !master_context_.dfly_session_id.empty());
+  ProactorBase* mythread = ProactorBase::me();
+  CHECK(mythread);
+
+  CHECK(sock_->IsOpen());
+  // sock_.reset(mythread->CreateSocket());
+  // RETURN_ON_ERR(sock_->Connect(master_context_.master_ep));
+  sync_fb_ = ::boost::fibers::fiber(&Replica::StableSyncDflyFb, this);
+
+  return std::error_code{};
+}
+
+void Replica::FullSyncDflyFb(SyncBlock* sb, string eof_token) {
+  DCHECK(leftover_buf_);
   SocketSource ss{sock_.get()};
-  io::PrefixSource ps{io_buf->InputBuffer(), &ss};
+  io::PrefixSource ps{leftover_buf_->InputBuffer(), &ss};
 
   RdbLoader loader(NULL);
+  loader.fullsyncb = [this, sb, ran = false]() mutable {
+    if (ran) return;
+    {
+      std::unique_lock lk(sb->mu_);
+      sb->flows_left--;
+      ran = true;
+    }
+    sb->cv_.notify_all();
+  };
   loader.Load(&ps);
 
+  // Try finding eof token.
+  io::PrefixSource chained_tail{loader.Leftover(), &ps};
   if (!eof_token.empty()) {
-    unique_ptr<uint8_t[]> buf(new uint8_t[eof_token.size()]);
-    // pass leftover data from the loader.
-    io::PrefixSource chained(loader.Leftover(), &ps);
-    VLOG(1) << "Before reading from chained stream";
-    io::Result<size_t> eof_res = chained.Read(io::MutableBytes{buf.get(), eof_token.size()});
-    if (!eof_res || *eof_res != eof_token.size()) {
+    unique_ptr<uint8_t[]> buf{new uint8_t[eof_token.size()]};
+
+    io::Result<size_t> res =
+        chained_tail.ReadAtLeast(io::MutableBytes{buf.get(), eof_token.size()}, eof_token.size());
+
+    if (!res || *res != eof_token.size()) {
       LOG(ERROR) << "Error finding eof token in the stream";
     }
+  }
 
-    // TODO - to compare tokens
+  // Keep laoder leftover.
+  io::Bytes unused = chained_tail.unused_prefix();
+  if (unused.size() > 0) {
+    leftover_buf_.reset(new base::IoBuf{unused.size()});
+    auto mut_bytes = leftover_buf_->AppendBuffer();
+    memcpy(mut_bytes.data(), unused.data(), unused.size());
+    leftover_buf_->CommitWrite(unused.size());
+  } else {
+    leftover_buf_.reset();
```

Comment:
I don't like it, but this is what I came up with. 

What if we instead allow move-stealing the loaders underlying `IoBuf`? This would save us from allocating a new leftover buffer. Btw, we can also re-use the previous by resetting it... Which would be `ConsumeInput(InputLen())` (and add a helper to helio for this)

Reply:
Both options are ok. it's an implementation detail and is not very interesting. performance impact is negligible too. So just choose what you prefer and do it. We can always change it in the future.